capsules/src/virtual_uart.rs

```diff
@@ -0,0 +1,212 @@
+        });
+        self.do_next_op();
+    }
+
+    fn receive_complete(
+        &self,
+        rx_buffer: &'static mut [u8],
+        rx_len: usize,
+        error: hil::uart::Error,
+    ) {
+        // Pass the data to each receiver that wants to receive messages.
+        self.devices.iter().for_each(|device| {
+            if device.receiver {
+                device.rx_buffer.take().map(|rxbuf| {
+                    let len = cmp::min(rx_len, device.rx_len.get());
+                    for i in 0..len {
+                        rxbuf[i] = rx_buffer[i];
+                    }
+                    device.receive_complete(rxbuf, len, error);
+                });
+            }
+        });
+        self.rx_buffer.replace(rx_buffer);
+    }
+}
+
+impl<'a> UartMux<'a> {
+    pub fn new(uart: &'a hil::uart::UART, rx_buffer: &'static mut [u8]) -> UartMux<'a> {
+        UartMux {
+            uart: uart,
+            devices: List::new(),
+            inflight: OptionalCell::empty(),
+            rx_buffer: TakeCell::new(rx_buffer),
+        }
+    }
+
+    fn do_next_op(&self) {
+        if self.inflight.is_none() {
+            let mnode = self.devices.iter().find(|node| node.operation.is_some());
+            mnode.map(|node| {
+                node.tx_buffer.take().map(|buf| {
+                    node.operation.map(move |op| match op {
+                        Operation::Transmit { len } => self.uart.transmit(buf, *len),
+                    });
+                });
+                node.operation.clear();
+                self.inflight.set(node);
+            });
+        }
+    }
+
+    /// If we are not currently listening, start listening.
+    fn start_receive(&self, rx_len: usize) {
+        self.rx_buffer.take().map(|rxbuf| {
+            let len = cmp::min(rx_len, rxbuf.len());
+            self.uart.receive(rxbuf, len);
+        });
+    }
+}
+
+#[derive(Copy, Clone, PartialEq)]
+enum Operation {
+    Transmit { len: usize },
+}
+
+pub struct UartDevice<'a> {
+    mux: &'a UartMux<'a>,
+    receiver: bool, // Whether or not to pass this UartDevice incoming messages.
+    tx_buffer: TakeCell<'static, [u8]>,
+    rx_buffer: TakeCell<'static, [u8]>,
+    rx_len: Cell<usize>,
+    operation: OptionalCell<Operation>,
+    next: ListLink<'a, UartDevice<'a>>,
+    client: OptionalCell<&'a hil::uart::Client>,
+}
+
+impl<'a> UartDevice<'a> {
+    pub const fn new(mux: &'a UartMux<'a>, receiver: bool) -> UartDevice<'a> {
+        UartDevice {
+            mux: mux,
+            receiver: receiver,
+            tx_buffer: TakeCell::empty(),
+            rx_buffer: TakeCell::empty(),
+            rx_len: Cell::new(0),
+            operation: OptionalCell::empty(),
+            next: ListLink::empty(),
+            client: OptionalCell::empty(),
+        }
+    }
+
+    /// Must be called right after `static_init!()`.
+    pub fn setup(&'a self) {
+        self.mux.devices.push_head(self);
+    }
```

Comment:
I don't think this needs addressing addressing immediately, but is this something that could eventually be handled by the newly-merged Component interface? e.g. this capsule somehow provides a Component that boards instantiate and `finalize` (which I suppose is still a function that must be called, but is at least one standard)?